Various improvements to `field_isomorphism()`.

skieffer opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #22736 
Fixes #22849 


#### Brief description of what is fixed or changed

Improvements to `field_isomorphism()`:
* Speed: try to minimize precision for `evalf`. (See Issue #15743.)
* Code safety: use a non-global mpmath context.
* Simplifications: cut steps that consider +/- cases.
* Add explanatory comments.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* polys
  * Speed up `field_isomorphism()`.
<!-- END RELEASE NOTES -->

skieffer committed:
Various improvements to `field_isomorphism()`.

* Speed: use evalf(a) == evalf(b) instead of evalf(a - b) == 0. (See Issue #15743.)
* Code safety: only alter `mp.dps` inside a `try-finally`.
* Simplifications: cut steps that consider +/- cases.
* Add explanatory comments.

oscarbenjamin commented on sympy/polys/numberfields/subfield.py:
```diff
@@ -93,40 +93,45 @@ def field_isomorphism_pslq(a, b):
+            mp.dps = n
+            coeffs = pslq(basis, maxcoeff=int(1e10), maxsteps=1000)
+        finally:
+            mp.dps = dps
```
It would be better to use a local context:
```python
In [30]: s2 = sqrt(2).evalf(200)

In [31]: from mpmath import mp

In [32]: ctx = mp.clone()

In [33]: ctx.dps = 100

In [34]: ctx.pslq([1, s2._mpf_, (s2**2)._mpf_])
Out[34]: [2, 0, -1]
```

skieffer replied:
Could we use `workdps`? As in,

```python
from mpmath import pslq, workdps
with workdps(n):
    coeffs = pslq(basis)
```

jksuom replied:
It seems to me that `workdps` will do the same `try - finally` operation under the hood. So it could be used. Maybe it would even be preferable.

oscarbenjamin replied:
Does `workdps` modify the `dps` for the global `mp` context? Is it thread-safe if another thread is using the `mp` context as well?

jksuom replied:
`workdps` is defined in `mpmath/__init__.py` as `mp.workdps` where `mp = MPContext()`. I guess we could use ` with MPContext().workdps(n):` to get a temporary local context.

oscarbenjamin replied:
Yes, that's better than using a clone of `mp` because a user might have altered the properties of the global `mp` context. I'm not sure we can use with and then use the global `mpmath.pslq` function though because I think that still gets its properties from the global context somehow. So I think it's:
```python
ctx = MPContext()
ctx.dps = n
coeffs = ctx.pslq(basis)
```
We shouldn't need to use `with` since this only modifies the properties of the throwaway `ctx`.

jksuom replied:
> We shouldn't need to use with

Agreed. I also came to that conclusion.

oscarbenjamin commented on sympy/polys/numberfields/subfield.py:
```diff
@@ -136,23 +141,27 @@ def field_isomorphism_pslq(a, b):
+            # Write c as polynomial in b
+            coeffs = c.to_sympy_list()
             d, terms = len(coeffs) - 1, []
-
             for i, coeff in enumerate(coeffs):
                 terms.append(coeff*b.root**(d - i))
-
             root = Add(*terms)
-
-            if (a.root - root).evalf(chop=True) == 0:
+            # Check whether we got the number a
+            if A is None:
+                A = a.root.evalf(chop=True)
+            if root.evalf(chop=True) == A:
```
Checking `a.evalf() == b.evalf()` is not equivalent to checking `(a - b).evalf()` if the numbers differ by an amount smaller then the default precision of `evalf`:
```python
In [40]: a = 1 + Rational(1, 10**20)

In [41]: b = S(1)

In [42]: a.evalf() == b.evalf()
Out[42]: True

In [43]: (a - b).evalf()
Out[43]: 1.00000000000000e-20
```

oscarbenjamin replied:
Maybe that's not a problem here though. We know that `a` and `b` are nonzero. Are they guaranteed to differ in sign or leading digit?

skieffer replied:
Not necessarily sign or leading digit, but I am thinking that an isolating interval approach could be the way to go.

oscarbenjamin replied:
Yes, that makes sense. What exactly does "conjugate" mean here? Does it mean that they have the same minimal polynomial?

skieffer replied:
Yup, exactly that.

So, yeah, for each linear factor, we've found _some_ root of the minimal polynomial of `a`, and we just want to know: is this `a` itself? It does sound like a job for isolating intervals.

And I guess computing adequate isolating intervals can be faster than working with evalf.

oscarbenjamin replied:
The root isolation code can definitely be improved so that refinement runs faster. The problem with evalf is that you need to isolate the root very closely but actually with a bounding interval if you know that `a` is a root of `p` then a bigger bounding interval is better than a smaller one:
```python
In [1]: r = RootOf(x**4 + x**3 + x**2 + x + 1, 0)

In [2]: r._get_interval()
Out[2]: (-1, -1/2) x (-1, -1/2)

In [3]: r.evalf()  # This is the slow part
Out[3]: -0.809016994374947 - 0.587785252292473⋅ⅈ

In [4]: r._get_interval()
Out[4]: (-932733090407273441/1152921504606846976, -1865466180814546881/2305843009213693952) x (-677670257458753271/1152921504606846976, -1355340514917506541/2305843009213693952)
```

skieffer replied:
Question about `evalf`: Is there a good way to get the desired number of decimal places strictly to the right of the decimal point, without computing a log?

Example:

Consider `a = 1000000 + sqrt(2)`, and suppose I want 10 places to the right of the decimal point.
Obviously this doesn't work:
```python
>>> a.evalf(10)
1000001.414
```
I can do it this way:
```python
>>> a.evalf(10 + 1 + int(log(a, 10)))
1000001.4142135624
```
but I'm wondering if there might be some ready-made solution for this.

oscarbenjamin replied:
I don't think that there is a ready-made solution although if you want this only for RootOf then there is probably a way to specify how small you want the bounding intervals to be.

Why exactly is this needed? There might be a different way to do this.

skieffer replied:
Well, here's what I'm trying to do: Say I have a `Poly` `f` with root isolating intervals `I1`, `I2`, ..., `In`.
I have two `Expr`s `r` and `s`, and I know `f(r) == 0 == f(s)`. I want to determine whether `r == s`.

But I don't want to compute `(r - s).evalf()`, because it's very slow. (#15743)

My idea is to let `D` be the largest denominator occurring in any endpoint of any of the intervals `Ik`.
Then `abs(r - s) < 1/D` would suffice to conclude that `r == s`.

But to check the inequality requires use of `evalf`. To what precision? My idea is to detemine `n` large
enough to guarantee that `abs(r - r.evalf(n)) < eps = 1/(3*D)`. (And likewise for `s`.)

Then `abs(r.evalf(n) - s.evalf(n)) < eps` will suffice to conclude `r == s`.

So I was thinking of giving `Poly` a method like this:

```python
    def root_comparison_tools(f):
        roots = f.all_roots(radicals=False)
        ivals = [r._get_interval() for r in roots]
        # Let D0 be max denom among all interval endpoints.
        D0 = int(max(i.max_denom for i in ivals))
        # We need 3 times that.
        D1 = 3*D0
        # Pick n large enough to guarantee 10^n > D1.
        n = math.ceil(D1.bit_length() / 3.3) + 2
        # Make a custom evalf function that guarantees rounding error < 10^(-n).
        evalf_ = lambda x: x.evalf(n + 1 + math.ceil(int(x).bit_length() / 3.3))
        # Define a threshold that guarantees two roots of f are the same.
        eps = S(1) / D1
        return eps, evalf_
```

and then the problem could be solved like this:

```python
eps, evalf_ = f.root_compartison_tools()
if abs(evalf_(r) - evalf_(s)) < eps:
    # roots r and s are the same...
else:
    # they're different roots...

```

oscarbenjamin replied:
So `(r - s).evalf()` is slow if `r` and `s` are equal because it ends up computing `r` and `s` to high precision in an attempt to find the significant digits of `r - s`. What's faster is computing the significant digits of something that is *nonzero*.

For each interval `Ik` (including the correct one) we expect that `abs(r - Ik.mid) - Di` does not equal zero and we can compute that to find whether it is positive or not e.g.:
```python
In [51]: p = x**4 + x + 2

In [52]: rs = [RootOf(p, n) for n in range(degree(p))]

In [53]: r1, r2, r3, r4 = roots(p)

In [54]: intervals = [r._get_interval() for r in rs]

In [55]: intervals[0]
Out[55]: (-1, -1/2) x (-1, -1/2)

In [56]: intervals[0].center
Out[56]: (mpq(-3,4), mpq(-3,4))

In [57]: intervals[0].dx
Out[57]: mpq(1,2)

In [58]: intervals[0].dy
Out[58]: mpq(1,2)

In [59]: r1.evalf()
Out[59]: -0.849848427793055 - 0.6542723031575⋅ⅈ

In [60]: def centre(i):
    ...:     a, b = i.center
    ...:     return a + b*I
    ...: 

In [61]: def radius(i):
    ...:     return i.dx + i.dy
    ...: 

In [62]: centre(intervals[0])
Out[62]: 
  3   3⋅ⅈ
- ─ - ───
  4    4 

In [63]: radius(intervals[0])
Out[63]: mpq(1,1)

In [64]: (abs(r1 - centre(intervals[0])) - radius(intervals[0])).evalf(2)
Out[64]: -0.86

In [65]: (abs(r1 - centre(intervals[1])) - radius(intervals[1])).evalf(2)
Out[65]: 0.41

In [66]: (abs(r1 - centre(intervals[2])) - radius(intervals[2])).evalf(2)
Out[66]: 0.71

In [67]: (abs(r1 - centre(intervals[3])) - radius(intervals[3])).evalf(2)
Out[67]: 1.5
```
Here `Out[64]` through to `Out[67]` proves that `r1` is the root bounded by `intervals[0]`. If we end up having multiple negatives here then it means we need to shrink the intervals and try again. This is analogous to the `choose_factor` method in minpoly: we prove a factor is the minpoly for `a` by proving that every other factor is *not* the minpoly. This is done by numerically showing that something is *not* zero which can be done reliably (unlike proving that something is zero).

Note that we don't actually need to do the `evalf(2)` explicitly here as it is precisely what `a > b` does for Expr:
```python
In [69]: abs(r1 - centre(intervals[3])) > radius(intervals[3])
Out[69]: True
```

Combining this for `r` and `s` an algorithm could be:
1. For each interval compute `(r - c)**2 < d**2`
2. Also compute `(s - c)**2 < d**2`
3. If either of `r` or `s` is definitely not in this interval discard the interval.
4. If we have one remaining interval then `r = s`.
5. If we have no remaining intervals then `r != s`.
6. If there are multiple intervals remaining, shrink them and go back to step 1.

That is:
```python
while True:
    intervals_keep = []
    for interval in intervals:
         c = centre(interval)
         d = radius(interval)
         # Should be a way to compute abs(z)**2 without intermediate square roots
         if abs(r - c)**2 < d**2 and abs(s - c)**2 < d**2:
             intervals_keep.append(interval)
    if not intervals_keep:
        return False
    elif len(intervals_keep) == 1:
        return True
    else:
        intervals = [shrink_interval(interval) for interval in intervals_keep]
```

It might seem inefficient to recalculate the `evalf` for each root but at least in the case of RootOf the work done by one `evalf` is carried over to the next. It would also be possible to wrap other kinds of expression with something that caches the results of `_eval_evalf`.

We know that `abs(r - Ik.mid) - Di` will not be zero for any real root because that would imply that `r` was rational. It might potentially be zero though for a Gaussian rational like `2 + 3*I` but it's unlikely.

skieffer replied:
Well, this is pretty cool. : )  I like the algorithm, and I like the principle that it's easy to check when something is _nonzero_, so work by process of elimination.

The note at the end about the outside possibility of an `evalf(0)` still happening is an interesting point. What if we used a transcendental radius, like `pi*(dx + dy)`. Then it's definitely different from `abs(r - c)` which is algebraic.

oscarbenjamin replied:
Yes, that makes sense. It could be `pi/3*(dx + dy)` to make the bound reasonably tight.

oscarbenjamin replied:
Or in fact we want `(dx + dy)/2` so `pi/6` should work.

skieffer replied:
Sounds good. I am also still interested to see whether the algorithm I described could work as well. I added a note correcting the way the denominator `D` is chosen.

Maybe it's worth coding them both up. Even if only one is used, I think they both contain valuable ideas that could be worth revisiting at some point.

Meanwhile, maybe the comprehensive move over to ball-radius arithmetic you mentioned [before](https://github.com/sympy/sympy/issues/15743#issuecomment-985706352) will eventually make it all irrelevant?

It does seem that a fundamental problem with `evalf()` is that it wants to give you so many digits of precision, when all you may really want is so much accuracy, i.e. to bound the rounding error by `10^(-n)`.

oscarbenjamin replied:
The algorithm I described above is quite similar to what `minpoly` does. Is it a bad idea to just compute `minpoly(r - s)`?

oscarbenjamin replied:
Oh, I suppose that had to `(r - s).evalf()` as well as all the factoring.

jksuom replied:
The minimum distance between roots of a polynomial `p` can be estimated as follows. The roots of the resultant of `p(x)` and `p(x + y)` are the differences `x - x'` for all pairs of roots `(x, x')` of `p`.
```
In [1]: p = x**4 + x + 2                                                        

In [2]: q = p.subs(x, x + y)                                                    

In [3]: r = resultant(p, q, x)                                                  

In [4]: r                                                                       
Out[4]: 
 16       12       10        8        6         4
y   + 16⋅y   + 26⋅y   - 448⋅y  + 432⋅y  + 2021⋅y
```
After removing the zero roots, a lower bound for `|1/y|` can be found by applying e.g. [Cauchy's bound](https://en.wikipedia.org/wiki/Geometrical_properties_of_polynomial_roots#Lagrange's_and_Cauchy's_bounds) to the reversed polynomial. In the above case we get `|1/y| <= 1 + 448/2021`. The result can be sharpened by noting that the resultant is an even polynomial (for each `x - x'`, also `x' - x` is a root) hence a polynomial in `y**2` so that Cauchy's bound is also valid for `|1/y**2|`.

skieffer replied:
Nice! So maybe the `Poly` class can have a method roughly like this:

```python
    def root_separation_lower_bound(f):
        """
        Compute a lower bound on the minimum distance between any two
        roots of this polynomial.
        """
        if f.is_multivariate:
            raise MultivariatePolynomialError(
                "must be a univariate polynomial")

        from sympy.functions.elementary.miscellaneous import sqrt

        x = f.gen
        y = Dummy('y')
        g = f.subs(x, x + y)
        h = Poly(resultant(f, g, x))
        # If f has roots ri, the roots of h are all the ri - rj.

        coeffs = h.rep.rep
        while coeffs[-1] == 0:
            coeffs.pop()
        # Use the Cauchy bound
        d = coeffs.pop()
        M = 1 + max(abs(n / d) for n in coeffs)
        # Since h is an even func., can improve bound with sqrt.
        return 1/sqrt(M)
```

oscarbenjamin replied:
Yes, and maybe there could also be a method just called `roots_bound` that just gives root bounds in general. Using `sqrt` here is tricky. We ideally want the bounds to be rational numbers. It might be better to have something that returns the square of the bound(s).

Also `rep.rep` should not be manipulated directly as it is a private internal attribute of the immutable `Poly`. There is `all_coeffs` to get the coefficients as `Expr`.

Also any methods like this should check the domain and whether the `Poly` is univariate.

jksuom replied:
Maybe `isqrt` could be used. Something like `Rational(isqrt(M.p*M.q) + 1, M.q)`.

skieffer replied:
Finally was able to get back to this...

On making the bounds rational: I've written it so that the bounds will be rational if the polynomial's coefficients are. Do you think this is good enough?

skieffer replied:
@jksuom Is it all right if I set you as first author of 9c91d52a6f? These are your ideas.

jksuom replied:
It is all right, though I wasn't expecting that.

jksuom commented on sympy/polys/numberfields/subfield.py:
```diff
@@ -93,40 +93,45 @@ def field_isomorphism_pslq(a, b):
         A = a.root.evalf(n)
         B = b.root.evalf(n)
 
-        basis = [1, B] + [ B**i for i in range(2, m) ] + [A]
```
It is true, of course, that if `-a` is in `Q(b)` then so is `a` but I suspect that this line was the reason for the unexplained tests with `+/-a`. Clearly, `-a` should be used here to avoid confusion.

skieffer replied:
True, I could see this having been overlooked, and then the cases further down were the result of trying to make it work.

oscarbenjamin commented:
The idea I had for improving the performance of RootOf refinement is basically the same as #19164 and is based on mpmath's interval arithmetic:
https://mpmath.org/doc/1.0.0/contexts.html#arbitrary-precision-interval-arithmetic-iv

Here's how it looks for refining bounds on an isolated real root:
```python
from sympy import *
from mpmath import MPIntervalContext


def qq_to_iv(num, iv):
    """Convert rational QQ to real MPF interval"""
    return iv.mpf(num.numerator) / iv.mpf(num.denominator)


def poly_to_iv(poly, iv):
    """Convert poly with QQ coefficients to MPF intervals"""
    return [qq_to_iv(coeff, iv) for coeff in poly]


def intersect_iv(X, Y, iv):
    a = max(X.a, Y.a)
    b = min(X.b, Y.b)
    if a <= b:
        return iv.mpf((a, b))
    else:
        return None


def is_proper_subset(X, Y):
    return X.a > Y.a and X.b < Y.b


def split_iv(X, iv):
    return iv.mpf((X.a, X.mid)), iv.mpf((X.mid, X.b))


def split_iv_zero(X, iv):
    assert 0 in X
    return iv.mpf((X.a, 0)), iv.mpf((0, X.b))


def horner(poly, x0):
    """Evaluate a poly

    >>> horner([1, 0, -1], 3)
    8
    """
    val = poly[0]
    for c in poly[1:]:
        val = c + val*x0
    return val


def interval_newton(poly, polydiff, X, prec=53):
    iv = MPIntervalContext()
    iv.dps = prec + 20

    poly_iv = poly_to_iv(poly, iv)
    polydiff_iv = poly_to_iv(polydiff, iv)
    X_iv = iv.mpf(X)

    queue = [X_iv]
    while queue:
        if len(queue) == 1:
            [Xj] = queue
            a, b = Xj.a, Xj.b
            if abs(a - b) * 2**(prec+3) < (abs(a) + abs(b)):
                return Float(Xj.mid, precision=prec)

        nextqueue = []

        for X1 in queue:
            x1 = X1.mid
            f = horner(poly_iv, x1)
            Fp = horner(polydiff_iv, X1)

            if 0 not in Fp:
                Z2 = x1 - f/Fp
                X2 = intersect_iv(Z2, X1, iv)

                if X2 is None:
                    pass # no roots
                elif is_proper_subset(Z2, X1):
                    # this interval is the one -> discard all others
                    nextqueue = [X2]
                    break
                elif 4*X2.delta < 3*X1.delta:
                    # converging -> continue
                    nextqueue.append(X2)
                else:
                    # insufficient progress -> split
                    nextqueue.extend(split_iv(X2, iv))

            else:
                # Split at the zero of the denominator
                Fp1, Fp2 = split_iv_zero(Fp, iv)
                X21 = intersect_iv(x1 - f/Fp1, X1, iv)
                X22 = intersect_iv(x1 - f/Fp2, X1, iv)
                X2s = [Xi for Xi in [X21, X22] if Xi is not None]
                nextqueue.extend(X2s)

        queue = nextqueue


def refine_root(p, a, b, dps=15):
    poly = Poly(p).rep.rep
    polydiff = Poly(p.diff()).rep.rep
    prec = dps_to_prec(dps)
    return interval_newton(poly, polydiff, (a, b), prec)


x = Symbol('x')
p = x**2 - 2
root = refine_root(p, 0, 1.5)
print(root)
```
With this you can refine a real root much faster than with the current `RootOf.evalf` method:
```python
In [3]: p = x**10 + x**4 - x**3 - 3

In [4]: %time ok = refine_root(p, 0, 2, 10000)
CPU times: user 304 ms, sys: 4 ms, total: 308 ms
Wall time: 307 ms

In [6]: %time ok2 = RootOf(p, 1).evalf(10000)
CPU times: user 1min 29s, sys: 16 ms, total: 1min 29s
Wall time: 1min 29s

In [9]: p = x**10 + x**8 + x**3 - 100

In [10]: %time ok = refine_root(p, 0, 2, 20)
CPU times: user 24 ms, sys: 0 ns, total: 24 ms
Wall time: 24.6 ms

In [11]: %time ok2 = RootOf(p, 1).evalf(20)
CPU times: user 48 ms, sys: 0 ns, total: 48 ms
Wall time: 47.4 ms

In [12]: ok == ok2
Out[12]: True
```
The code above only works for real roots but it can be extended to real systems of equations and therefore can also find the complex roots of a polynomial.

skieffer committed:
User non-global MP context to adjust precision.

skieffer committed:
Add `Poly.root_bounds()` method.

jksuom committed:
Add `Poly.root_separation_lower_bound()` method.

Co-authored-by: Steve Kieffer <sk@skieffer.info>

skieffer committed:
core/evalf: Add `fastlog10_for_expr()` function.

skieffer committed:
Add `Poly.root_comparison_tools()` method.

skieffer committed:
Use `Poly.root_comparison_tools()`

We make use of this in:
* `field_isomorphism_factor()`
* `AlgebraicNumber.to_root()`
to make these functions faster and simpler.

skieffer commented:
Some good news: Test `sympy.polys.numberfields.tests.test_subfield.test_issue_22736()` was taking over 9 seconds before these changes, and now takes only about 1/4 of a second!

skieffer committed:
Try to pass flake tests.

oscarbenjamin commented on sympy/core/evalf.py:
```diff
@@ -149,6 +149,20 @@ def fastlog(x: Optional[MPF_TUP]) -> tUnion[int, Any]:
+    # up for it by adding 1 at the end.
+    a = abs(v.evalf(1))
+    b = int(math.ceil(a)).bit_length()
+    c = math.ceil(b / LG10)
```
It looks like `math.ceil(abs(v))` will call `evalf` implicitly:
```python
In [9]: x = 10**100*pi

In [10]: math.ceil(x).bit_length()
Out[10]: 334
```
Why do we need `fastlog10` instead of `fastlog` above which uses base 2?

skieffer replied:
The `.evalf(1)` makes it faster, and is good enough for the approximation we want here.

```python
f = cyclotomic_poly(5)
v = CRootOf(f, 0)
timeit.timeit('v._reset(); math.ceil(abs(v))', number=10, globals=globals())
Out[7]: 10.512659683999999
timeit.timeit('v._reset(); math.ceil(abs(v.evalf(1)))', number=10, globals=globals())
Out[8]: 0.9664351640000035
```

skieffer replied:
On `fastlog10` versus `fastlog`: I discuss below, where it comes up again.

oscarbenjamin commented on sympy/polys/numberfields/subfield.py:
```diff
@@ -88,45 +88,47 @@ def field_isomorphism_pslq(a, b):
-        coeffs = pslq(basis, maxcoeff=int(1e10), maxsteps=1000)
-        mp.dps = dps
+        ctx.dps = n
+        coeffs = ctx.pslq(basis, maxcoeff=int(1e10), maxsteps=1000)
```
I would write `int(1e10)` as `10**10`

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3773,6 +3774,178 @@ def nth_power_roots_poly(f, n):
+        lb, _ = h.root_bounds()
+
+        # Since h is an even function, can improve bound with square root.
+        if lb.is_rational:
```
This should probably be `is_Rational` rather than `is_rational`. Note the difference:
```python
In [4]: a = S(3)

In [5]: b = Symbol('b', rational=True)

In [6]: a
Out[6]: 3

In [7]: b
Out[7]: b

In [8]: a.is_rational
Out[8]: True

In [9]: a.is_Rational
Out[9]: True

In [10]: b.is_rational
Out[10]: True

In [11]: b.is_Rational
Out[11]: False
```
The test `is_Rational` is structural: is this an explicit rational number. The test `is_rational` is a *semantic* query through the assumptions system:
https://docs.sympy.org/latest/guides/assumptions.html

The polys module is mostly concerned with structural considerations rather than semantic in this sense.

skieffer replied:
Ah, thanks for the explanation. i was wondering about this.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3773,6 +3774,178 @@ def nth_power_roots_poly(f, n):
+            return Rational(isqrt(lb.p*lb.q) + 1, lb.q)
+        else:
+            from sympy.functions.elementary.miscellaneous import sqrt
+            return sqrt(lb)
```
If the function just returns the square of the bound then the shenanigans here are not needed and the return value will always be within the field generated by the coefficients.

skieffer replied:
In other words, you're saying that sure the square root sharpens the bound, but who cares. Maybe it's not worth the extra calculation. I suppose that may be true in many cases.

Maybe the method should take a keyword arg, `take_sqrt=False` so those who care can activate it.

oscarbenjamin replied:
The sharp bound is good. It's just that I can have a function that returns the bound or a function that returns the square of the bound and each is equally usable with the same sharpness:
```python
if abs(value) > bound():
    # do stuff

if value ** 2 > bound_squared():
    # do the same stuff
```
The difference between these two cases is that in the latter I can do all calculations within the same field e.g. if `bound_squared()` is guaranteed to return a rational number and value is a rational number then the check `value ** 2 > bound_squared()` is pure rational arithmetic and does not require computing in an extension field or using floating point.

In the complex case even `abs(value)` involves a square root but `abs(value)**2` is `re(value)**2 + im(value)**2` so if both parts are rational then the result is rational.

skieffer replied:
Okay, those are some excellent points. I suppose as long as the method is called something like `root_separation_lower_bound_squared()`, users will know what they're getting. Also the examples in the method's docstring can show the usage patterns you just described.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3773,6 +3774,178 @@ def nth_power_roots_poly(f, n):
+            coeffs.pop()
+        lc, tc = coeffs[0], coeffs[-1]
+        ub = 1 + max(abs(n / lc) for n in coeffs[1:])
+        lb_inv = 1 + max(abs(n / tc) for n in coeffs[:-1])
```
It does seem somewhat backwards to use Expr here. Can we not convert to a field and compute this with domain elements e.g.:
```python
In [24]: p = Poly(x**2 + 2)

In [25]: p = p.to_field()

In [26]: coeffs = p.rep.rep

In [27]: result = 1 + max(abs(n/coeffs[0]) for n in coeffs[1:])

In [28]: result
Out[28]: mpq(3,1)

In [29]: K.to_sympy(result)
Out[29]: 3
```

skieffer replied:
Actually running into an interesting issue here, depending on the domain of the `Poly`, since `abs()` is not supported for domain elements of types `ANP` or `GuassianElement`.

What do you suppose is the right way to handle this? Should these domain types define an `__abs__()` method?

For example, I tried giving `GaussianRational` this method:

```python
    def __abs__(self):
        from sympy.polys.domains.complexfield import CC
        z = CC.from_GaussianRationalField(self, self.base)
        return abs(z)
```

which does solve the immediate problem. But unfortunately importing `CC` at the top of the module causes a cyclic import, and importing it here seems a bit ugly. Is there a better way?

oscarbenjamin replied:
I still think that we should try to avoid introducing square roots if it's possible.

Maybe it would be better to give a domain method like `K.abs2` which can compute `abs(z)**2` since I think it should always be possible to compute that within the same ring:
```python
def abs2(K, z):
    return z.x**2 + z.y**2
```
For ANP it could be done if there was a way to compute complex conjugates...

skieffer replied:
Yeah, if it's important to stay in the Poly's domain, then I think there will be trouble with `ANP` (e.g. the fields generated by the cube roots of 2 -- the universal counterexample!)....

Also if `Poly.root_bounds()` returns squared bounds, then the `Poly.root_separation_...` method will return the 4th power of the bound, and I feel like it's getting out of hand...

I'm thinking I'd really like to retreat a bit here. How about if `Poly.root_bounds()` only supports polys over `QQ` and `ZZ`. That's all that was originally called for. I was trying to support all domains contained in `CC` because it seemed easy, but it no longer does. If someone needs that at some point, let them open an issue at that time.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3773,6 +3774,178 @@ def nth_power_roots_poly(f, n):
+        # Make a custom evalf function that guarantees rounding error < 10^(-n).
+        # This means we want more than n places to the right of the decimal
+        # point. `L(x)` gets us past any places to the left of the point.
+        evalf_ = lambda x: x.evalf(L(x) + n + 1)
```
Is this the reason for needing fastlog10?

Maybe `dps_to_prec` could be used here to work in binary precision.

skieffer replied:
Yes, this was part of the reason. Also I wanted a function that would accept any `Expr`, instead of the `MPF_TUP` that's required by `fastlog`.

Maybe, if we try to work with domain elements instead of `Expr`, as you discuss above, this could be reconsidered.

oscarbenjamin replied:
The `evalf` function in the `sympy.core.evalf` module takes a binary precision but otherwise does more or less the same as the `Expr.evalf` method.

The mpf tuple is easy to obtain:
```python
In [1]: pi.evalf(1)._mpf_
Out[1]: (0, mpz(25), -3, 5)
```

jksuom commented on sympy/polys/numberfields/subfield.py:
```diff
@@ -88,45 +88,47 @@ def field_isomorphism_pslq(a, b):
     g = b.minpoly.replace(f.gen)
 
     n, m, prev = 100, b.minpoly.degree(), None
+    ctx = MPContext()
```
Local context is a good idea but it would probably be better to use `PythonMPContext`. There are unresolved issues in the implementation of `MPContext` (see #19153). There is a single static `RealElement` (and `ComplexElement`) class whose `context` attribute is patched to point to the new `MPContext` class each time one is created. So all computations in existing `RR` and `CC` classes will always have the precision set in the last `MPContext`.

skieffer replied:
Is this a problem with the `MPContext` imported directly from `mpmath`, or just with the one defined [in SymPy](https://github.com/sympy/sympy/blob/d822fcba181155b85ff2b29fe525adbafb22b448/sympy/polys/domains/mpelements.py#L45)?

I tried using `PythonMPContext`, but it has no `.pslq` attribute.

oscarbenjamin replied:
In mpmath it's added through the IdentificationMethods mixin class:
https://github.com/fredrik-johansson/mpmath/blob/c11db84b3237bd8fc6721f5a0c5d7c0c98a24dc1/mpmath/identification.py#L837
That class is then a superclass for StandardBaseContext. Finally MPContext subclasses both BaseMPContext and StandardBaseContext:
https://github.com/fredrik-johansson/mpmath/blob/c11db84b3237bd8fc6721f5a0c5d7c0c98a24dc1/mpmath/ctx_mp.py#L58
No idea why it's all done in such a complicated way.

skieffer replied:
I think we might be okay. Calling `field_isomorphism_pslq()` does not appear to be altering the context of a `RealElement`:

```python
In [2]: from sympy import *
   ...: from sympy.polys.numberfields.subfield import field_isomorphism_pslq
   ...: a = RR(2)
   ...: print(a.context.prec)
   ...: b = AlgebraicNumber(sqrt(2))
   ...: c = AlgebraicNumber(sqrt(2) + sqrt(3))
   ...: print(field_isomorphism_pslq(b, c))
   ...: print(a.context.prec)
53
[1/2, 0, -9/2, 0]
53
```

skieffer commented on sympy/polys/polytools.py:
```diff
@@ -3773,6 +3774,178 @@ def nth_power_roots_poly(f, n):
+        Produce a threshold ``eps`` and custom evalf function ``evalf_`` such
+        that two roots ``a``, ``b`` of this polynomial must be equal if
+        ``abs(evalf_(a) - evalf_(b)) < eps``.
+
```
I ran some tests, and

```python
A = evalf_(a)
for b in roots:
    abs(A - evalf_(b)) < eps
```

did not appear to be any faster than

```python
for b in roots:
    abs(evalf_(a) - evalf_(b)) < eps
```

(Due to caching, I suppose?)

So maybe this method can be simplified. It could be called `root_equality_test()`, and could simply return a binary function like

```python
lambda a, b: abs(evalf_(a) - evalf_(b)) < eps
```

(or one that incorporates the squared bound idea).

oscarbenjamin replied:
A RootOf caches its internal interval so repeated calls to evalf are immediate:
```python
In [3]: r = RootOf(x**3 + x + 1, 0)

In [4]: r
Out[4]: 
       ⎛ 3           ⎞
CRootOf⎝x  + x + 1, 0⎠

In [5]: %time r.evalf()
CPU times: user 20 ms, sys: 0 ns, total: 20 ms
Wall time: 19 ms
Out[5]: -0.682327803828019

In [6]: %time r.evalf()
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 878 µs
Out[6]: -0.682327803828019
```
Internally `evalf` will try increasing the precision on subterms until it considers that the overall result has the accuracy it wants. So when you have `(a - b).evalf()` if `a = b` then `evalf` will increase the precision again and again until it hits the limit. Each time it does that RootOf needs to shrink its interval so we end up needing a much smaller interval then we would for `a.evalf()` or even `(a.evalf() - b).evalf()`. This is the part that is slowest.

These caches will affect your timings if you do everything in the same process. Instead make a complete Python script and time the whole operation:
```python
from sympy import *

x = Symbol('x')
p = x**4 + x**3 + x**2 + x + 1
rs = roots(p)
r3 = RootOf(p, 3)

def f():
    return [abs(r3 - r).evalf() for r in rs]

def g():
    r3e = r3.evalf()
    return [abs(r3e - r).evalf() for r in rs]

if True:
    f()
else:
    g()
```
That script will call `f` and the timing we get is:
```console
$ time python r.py

real	0m36.530s
user	0m36.496s
sys	0m0.024s
```
Now instead let's call `g()` (change the True to False):
```console
$ time python r.py

real	0m1.505s
user	0m1.464s
sys	0m0.040s
```
So `f` is a lot slower than `g` and this slowness is absolutely on a scale that is noticeable to a user.

If you look at the outputs of `f` and `g` you can also see that they don't get the same result:
```python
In [3]: f()
Out[3]: 
[1.90211303259031 - 0.e-25⋅ⅈ, 1.17557050458495 - 0.e-25⋅ⅈ, 1.90211303259031 + 0.e-22⋅ⅈ, 2.952700857
5875e-64 - 1.25078415499626e-63⋅ⅈ]

In [4]: g()
Out[4]: [1.90211303259031, 1.17557050458495, 1.90211303259031, 7.81510642993629e-9]
```
The absolute difference computed by `f` is smaller by 50 orders of magnitude.

Personally I think that both `f` and `g` are too slow. What needs to be improved for performance here is the algorithm that shrinks the intervals.

skieffer replied:
Thanks. I didn't realize separate processes should be called in order to get a good comparison.

In the two forms I want to consider, the `evalf` always happens before the subtraction.
So, maybe the fact about caching already settles the question.
But just to be sure I add functions `h` and `k` to the script:

```python
import sys
from sympy import *

x = Symbol('x')
p = x**4 + x**3 + x**2 + x + 1
rs = roots(p)
r3 = RootOf(p, 3)

def f():
    return [abs(r3 - r).evalf() for r in rs]

def g():
    r3e = r3.evalf()
    return [abs(r3e - r).evalf() for r in rs]

def h():
    return [abs(r3.evalf() - r.evalf()) for r in rs]

def k():
    r3e = r3.evalf()
    return [abs(r3e - r.evalf()) for r in rs]

funcs = [f, g, h, k]

F = funcs[int(sys.argv[1])]
F()
```

and we get:

```bash
$ time python r.py 0

real    0m18.880s
user    0m18.381s
sys 0m0.153s

$ time python r.py 1

real    0m1.000s
user    0m0.937s
sys 0m0.056s

$ time python r.py 2

real    0m0.960s
user    0m0.891s
sys 0m0.052s

$ time python r.py 3

real    0m0.974s
user    0m0.895s
sys 0m0.057s
```

So I do think the `root_comparison_tools()` method I wrote could be simplified. I made it return the pair `(eps, evalf_)` because I imagined it might offer users an important performance improvement to let them use form `k()`. But, if in fact form `h()` is just as good, then it can be a simpler method that returns a single test function

```python
lambda a, b: abs(evalf_(a) - evalf_(b)) < eps
```

> Personally I think that both f and g are too slow. What needs to be improved for performance here is the algorithm that shrinks the intervals.

To me it depends on the application. For certain web apps, clicking and waiting one second to see a result might be just fine, whereas waiting 10s starts to feel unresponsive.

skieffer committed:
Use int power of 10.

skieffer committed:
Refactor: introduce `quad_to_mpmath()` function.

skieffer committed:
Add `evalf_with_bounded_error()` function.

We also cut the experimental `fastlog10_for_expr()` function.

skieffer committed:
Rework `Poly.root_bounds()`: limit to `QQ` and `ZZ`

skieffer committed:
Update `Poly` root separation bound method.

* Return the square of the bound, instead of the bound itself.
  This allows us to keep the bounds in the domain of the coeffs.
* Limit domains to `QQ` and `ZZ`.

skieffer committed:
Introduce `Poly.root_equality_test()`

This replaces the `Poly.root_comparison_tools()` method.

skieffer committed:
Use `Poly.root_equality_test()`.

skieffer commented on sympy/core/evalf.py:
```diff
@@ -1750,3 +1753,56 @@ def N(x, n=15, **options):
+
+    """
+    if eps is not None:
+        if not eps.is_real or not eps > 0:
```
Was wondering if there's an `is_Real` versus `is_real` consideration here. But `is_Real` doesn't seem to be a thing?

skieffer replied:
Now I see that a `mypy` test is failing due to an `Expr` not having an `is_real` attribute. So, what's the right way to check whether an arbitrary `Expr` represents a real number?

oscarbenjamin replied:
Every Expr has an `is_real` attribute so maybe mypy just needs to be told that. Something like this:
https://github.com/sympy/sympy/blob/d822fcba181155b85ff2b29fe525adbafb22b448/sympy/core/basic.py#L84-L85

In fact every Basic has `is_real` but I think that really it should only be defined for Expr (#19474).

Note though that `is_real` is a *semantic* check (e.g. `Symbol('x', real=True).is_real`) which is not what is wanted here. Here I think that what is wanted is for `eps` to be an explicit elementary number like `Float/Rational`.

skieffer replied:
Okay, thanks. I've updated the type checking.

skieffer committed:
Try to pass flake tests.

skieffer commented:
## Summary of recent changes

* Dumped the `fastlog10_for_expr()` function I had written in `core.evalf`. Instead, we now have `evalf_with_bounded_error()`. I couldn't find any existing function in either `core.evalf` or `mpmath` that would let you evalf up until the _absolute error_ was less than a given bound. So that's what this function does.

* The new `Poly` methods:

    - `root_bounds()` -- and as a consequence, the other two methods as well -- now only accepts polys over `QQ` or `ZZ`.
    - `root_separation_lower_bound_squared()` does just that: it returns the _square_ of the Cauchy lower bound on the
       separation between distinct roots of this polynomial.
    - `root_equality_test()` returns a binary function `eq(a, b)` such that, if `a` and `b` are complex numbers known to be roots of this polynomial, then `eq(a, b)` is `True` if and only if they are the _same_ root.

* `field_isomorphism_factor()` now uses `root_equality_test()` and as a consequence is a lot faster!

skieffer committed:
Repair type check in `evalf_with_bounded_error()`.

skieffer committed:
Simplify `evalf_with_bounded_error()`.

Don't actually need special handling for the case `x == 0`.

skieffer commented:
The simplification I just introduced in `evalf_with_bounded_error()` happened to expose a really surprising issue. It should be solved by #22850.

skieffer committed:
Solve issue with low-precision evalf of Alg Num.

skieffer commented:
This PR now includes a solution to #22849 as well as #22736, and I think it's ready from my point of view.

For reviewers: There are three different things happening in the file `sympy/core/evalf.py`:

* Introduces `quad_to_mpmath()` to do some refactoring of pre-existing code.
* Introduces `evalf_with_bounded_error()`.
* Adds the `AlgebraicNumber` class to the evalf table.

I think the changes in all the other files are more or less straightforward.

oscarbenjamin commented on sympy/core/evalf.py:
```diff
@@ -1750,3 +1758,54 @@ def N(x, n=15, **options):
+
+    """
+    if eps is not None:
+        if not (eps.is_Rational or eps.is_Float) or not eps > 0:
```
Maybe `eps` should be sympified.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3773,6 +3775,174 @@ def nth_power_roots_poly(f, n):
+    def root_bounds(f):
+        """
+        Compute lower and upper bounds on the magnitude of the nonzero roots of
+        this polynomial.
```
I think it would be good to clarify in this docstring somewhere that we are thinking about potentially non-real roots.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3773,6 +3775,174 @@ def nth_power_roots_poly(f, n):
+        # |a - ev(a)| < eps and |b - ev(b)| < eps.
+        # So if |ev(a) - ev(b)|**2 < eps**2, then
+        # |ev(a) - ev(b)| < eps, hence |a - b| < 3*eps = delta.
+        def eq_test(a, b):
+            A, B = ev(a), ev(b)
+            return (A.real - B.real)**2 + (A.imag - B.imag)**2 < eps_sq
+
+        return eq_test
```
Returning the function like this seems like an odd interface

skieffer replied:
I guess since the `Poly` is considered immutable it could just store the test function as an attribute, and return only the True/False answer to the user.

skieffer replied:
But now that I take a closer look, I see that the `Poly` class is actually really minimal in its instance attributes. It only has `rep` and `gens`, and it declares these as its slots `__slots__ = ('rep', 'gens')`. I'm not sure I want to disturb this....

oscarbenjamin replied:
Is there much value in storing anything?

oscarbenjamin replied:
If there is then it could be possible to have a cached function that returns the `eq_test` but still expose an interface to the user that looks more natural like `p.same_roots(r1, r2)`.

skieffer replied:
> Is there much value in storing anything?

I was running some timing tests in order to try to answer this, when I realized it's not scaling as well as I had hoped. I should have done these scaling tests much earlier!

I want to try `dup_zz_mignotte_bound` and see if it will do better. ([Mignotte bound](https://en.wikipedia.org/wiki/Geometrical_properties_of_polynomial_roots#Root_separation))

skieffer replied:
Couple of findings: (a) `dup_zz_mignotte_bound` is actually a different Mignotte bound, but anyway (b) the Mignotte bound I meant isn't doing much better than the bound we already have.

Might have to think some more...

jksuom replied:
Mignotte bound is currently misnamed.

oscarbenjamin replied:
What exactly is slow?

skieffer replied:
> What exactly is slow?

I'm looking at cyclotomic polynomials `phi_n`. I didn't anticipate how fast the coefficients of `Res(phi_n(x), phi_n(x + y))` would grow with `n`.

Of course for cyclotomic polys the geometry is easy, so better separation bounds could be used.

Anyway, the comparison with Mignotte is interesting. Our bound starts out several orders of magnitude better, but then gets worse. As for the run time, we're slowing down dramatically but Mignotte barely changes at all.


```python
f = lambda n: Poly(cyclotomic_poly(prime(n)))

def mignotte(f):
    n = f.degree()
    D = f.discriminant()
    eucl_norm = sqrt( sum( [cf**2 for cf in f.all_coeffs()] ) )
    denom = sqrt(n)**(n+1) * eucl_norm**(n-1)
    num = sqrt(3*abs(D))
    return num/denom

for n in range(2, 11):
    t = time.time()
    b1 = sqrt(f(n).root_separation_lower_bound_squared()).evalf()
    t1 = time.time() - t

    t = time.time()
    b2 = mignotte(f(n)).evalf()
    t2 = time.time() - t
    print(f'{prime(n)}:\n{b1}, {b2}')
    print(f'{t1:.5f}, {t2:.5f}\n')

3:
0.866025403784439, 0.612372435695794
0.00404, 0.00108

5:
0.845154254728517, 0.0541265877365274
0.00793, 0.00131

7:
0.509175077217316, 0.00327364250549328
0.01933, 0.00160

11:
0.00671827668354308, 0.00000547722557505166
0.11216, 0.00210

13:
0.000155324265551629, 1.67448988340192E-7
0.28852, 0.00230

17:
2.09849711164693E-9, 1.00818626091867E-10
1.15808, 0.00292

19:
1.82097216957742E-12, 2.05812538301822E-12
2.17408, 0.00345

23:
2.41307054558417E-20, 6.31974853510620E-16
6.41855, 0.00444

29:
4.75406277160334E-36, 1.79791507099308E-21
24.15388, 0.00471

```

skieffer commented:
"Root identification" is now looking to me like a whole subject of its own, and we should wrap up this PR without trying to settle that subject too. This PR contributes a little to that subject, and does several other things as well.


## Wrapping up

I think the last question we were considering was whether it was important for `Poly.root_equality_test()` to cache the test function. With the time costs involved in other parts of the process, I don't think this matters.


## Thoughts for future work on Root Identification

* References
    - [Mignotte 1982](http://people.dm.unipi.it/gianni/AC-EAG/Mignotte.pdf)
    - [Yap 2000](http://160592857366.free.fr/joe/ebooks/ShareData/Fundamental%20Problems%20in%20Algorithmic%20Algebra%201993%20Yap.pdf), Lecture VI, Sec. 7

* I am convinced of the necessity to improve the performance of `RootOf` refinement, which @oscarbenjamin has pointed to several times.

* There are several ways to compute lower bounds on root separations.

    - the Mignotte bound
    - the resultant method, which outperforms Mignotte up to a point, but then not
    - special cases, e.g. separation of `2*sin(pi/n)` for `cyclotomic_poly(n)`

* I still think the root intervals themselves can produce a better separation bound, albeit with a time tradeoff. E.g. on `f = cyclotomic_poly(29)`:
    - Mignotte is done in under 5ms, but the lower bound is `1.8E-21`, and refining roots of `f` to 21 places takes 27.4s.
    - Calling `f.intervals()` takes 1.6s, but the lower bound is `0.125`, and the root refinement takes 8.12s. Still not good, but better overall.

skieffer committed:
`sympify` `eps` in `evalf_with_bounded_error()`.

skieffer committed:
Clarify docstring in `Poly.root_bounds()`.

skieffer committed:
Change interface of `Poly.root_equality_test()`

Now accepts the roots `a`, `b` to be tested, and returns boolean.

oscarbenjamin commented:
The parts of this PR that I'm not sure about is where there are changes to public interfaces. Otherwise I would be happy to just merge as is with future improvements deferred for now.

One examples is the `evalf_with_bounded_error` function. It isn't clear if this is a public or a private function: are we committing to supporting that function as used by users of SymPy in future?

From an API design perspective I would have thought that it would make more sense to make bounded error an option to `evalf` so that the interface is more like:
```
expr.evalf(20, absolute_bound=50)
```
Which would mean something like "compute the expression with a relative error of 1e-20 and an absolute error of 1e-50". Perhaps instead the bound could be given as an explicit number like `1e-50`.

In any case it's clear if you look at the `.evalf` code that implementing absolute error bounding would require deep changes throughout the stack so it's not something I propose to do. I just think that before changing public API it's important to consider what the ideal API is rather than introduce something temporary.

Also the methods that are added to Poly include `root_equality_test` which has an interface that seems unintuitive to me. Also `root_bounds` really just computes the Cauchy bounds so I wonder if it would make sense to call it `cauchy_bounds`.

Another question from a codebase design perspective is if it really makes sense to have a method like `cauchy_bounds` as something like `dup_cauchy_bounds` so that it can be used at the lower-levels without needing to wrap things up in a Poly.

skieffer commented:
(1)

> From an API design perspective I would have thought that it would make more sense to make bounded error an option to `evalf`

I agree, that is probably the right design. For now do you think it would be enough to rename `evalf_with_bounded_error` with a leading underscore to make it clear it's not public?

(2)

> `root_equality_test` which has an interface that seems unintuitive to me.

Yes, I no longer think that caching the test function is important, so this can be more like,

```python

    def root_equality_test(f, a, b):
        ...
        return boolean saying whether a and b are equal
```

What about the name? `root_equality_test()`? `test_root_equality()`? `roots_are_equal()`?

(3)

> have a method like `cauchy_bounds` as something like `dup_cauchy_bounds` so that it can be used at the lower-levels without needing to wrap things up in a Poly

Do you have a sense for the module where `dup_cauchy_bounds()` belongs? I'd propose `polys.rootisolation`, which is filled with many `dup_...` methods on the topic of root geometry. In fact, it already has `dup_root_upper_bound()` (which appears to apply only to real roots).

(4)

> `root_bounds` really just computes the Cauchy bounds so I wonder if it would make sense to call it `cauchy_bounds`

In a similar sense, I wonder if we should be cautious about what we're offering with `Poly.root_separation_lower_bound_squared()`. Maybe for now it would be better just to offer the Mignotte bound, since it's a known thing, and save this idea for future work.

If so, then maybe it should be `dup_mignotte_sep_bound()`, again in the `polys.rootisolation` module?

oscarbenjamin commented:
> For now do you think it would be enough to rename `evalf_with_bounded_error` with a leading underscore to make it clear it's not public?

Yes.

> What about the name? `root_equality_test()`? `test_root_equality()`? `roots_are_equal()`?

What about `same_root` or `is_same_root/are_same_root`?

> I'd propose `polys.rootisolation`

I agree.

> If so, then maybe it should be `dup_mignotte_sep_bound()`, again in the `polys.rootisolation` module?

Yes maybe. Not sure...

I'm less concerned about adding "public" `dup_` functions than public Poly methods. There are hundreds of `dup_` functions and it's not hard to just leave one there once it's added.

skieffer committed:
Add `dup_...` funcs for Cauchy bounds.

skieffer committed:
Add `l2_norm_squared` methods.

skieffer committed:
Add squared Mignotte sep bound function.

skieffer committed:
Add new methods to `DMP` class, corresp. to new funcs.

skieffer committed:
Give `Poly` a `same_root()` method.

We cut the three experimental `Poly` methods from before:
* `root_bounds()` is now replaced by the `dup_cauchy...` functions
* `root_separation_lower_bound_squared()` is replaced by `dup_mignotte_sep_bound_squared()`
* `root_equality_test()` is replaced by `same_root()`.

skieffer committed:
Mark `evalf_with_bounded_error()` as private.

skieffer committed:
Move `sympify` call to make mypy happy.

skieffer committed:
Repair docstring.

skieffer commented:
Okay, I have addressed the last four things discussed.

For a root separation bound, I implemented Mignotte, as `dup_mignotte_sep_bound_squared()` in `polys.rootisolation`.

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3773,6 +3775,57 @@ def nth_power_roots_poly(f, n):
+        r, _, _, _ = evalf(1/eps_sq, 1, {})
+        n = fastlog(r)
+        # Then 2^n > 1/eps**2.
+        m = math.ceil(n/2)
```
Maybe I'm overthinking this but at first glance this seems an unnecessary use of machine precision floating point...

oscarbenjamin replied:
If `n` is guaranteed to be an an integer then I guess this would work:
```
m = (n // 2) + (n % 2)
```
(Add zero if n is even or 1 if n is odd)

skieffer replied:
Interesting. Would not have thought about the efficiency here. Will change it.

oscarbenjamin replied:
It's not so much efficiency as correctness:
```python
In [46]: e = 2 ** 53 + 1

In [47]: math.ceil(e / 2)
Out[47]: 4503599627370496

In [48]: (e // 2) + (e % 2)
Out[48]: 4503599627370497
```
That's probably way outside the range of what is possible here but in general we should avoid trusting floating point calculations unless necessary.

oscarbenjamin commented on sympy/polys/tests/test_densearith.py:
```diff
@@ -970,6 +971,19 @@ def test_dmp_l1_norm():
+    assert dmp_l2_norm_squared([[[]]], 2, ZZ) == 0
+    assert dmp_l2_norm_squared([[[1]]], 2, ZZ) == 1
+
+    assert dmp_l2_norm_squared(f_0, 2, ZZ) == 111
```
Why the blank line here?

oscarbenjamin commented:
I had a couple of minor comments but otherwise I think this looks good.

jksuom commented on sympy/polys/densearith.py:
```diff
@@ -1786,6 +1786,48 @@ def dmp_l1_norm(f, u, K):
+    if not f:
+        return K.zero
+    else:
+        return sum([ coeff**2 for coeff in f ])
```
Maybe something like `return sum([coeff**2 for coeff in f], K.zero)` would suffice.

skieffer replied:
I think so. I wrote it this way only to follow suit with the similar methods, like `dup_l1_norm`.

skieffer committed:
Stick to exact operations.

skieffer committed:
Delete blank line.

skieffer committed:
Simplify code.

skieffer committed:
Remove unused import.

skieffer commented:
Thanks @oscarbenjamin and @jksuom for reviewing! I've addressed your most recent comments. If there's nothing else, I'll merge soon.

oscarbenjamin commented:
Looks good to me!

jksuom commented:
Looks good to me too!
